wit/incoming-handler.wit

```diff
@@ -0,0 +1,24 @@
+// The `wasi:http/incoming-handler` interface is meant to be exported by
+// components and called by the host in response to a new incoming HTTP
+// response.
+//
+//   NOTE: in Preview3, this interface will be merged with
+//   `wasi:http/outgoing-handler` into a single `wasi:http/handler` interface
+//   that takes a `request` parameter and returns a `response` result.
+//
+default interface incoming-handler {
+  use pkg.types.{incoming-request, response-outparam}
+
+  // The `handle` function takes an outparam instead of returning its response
+  // so that the component may stream its response while streaming any other
+  // request or response bodies. The callee MUST write a response to the
+  // `response-out` and then finish the response before returning. The `handle`
+  // function is allowed to continue execution after finishing the response's
+  // output stream. While this post-response execution is taken off the
+  // critical path, since there is no return value, there is no way to report
+  // its success or failure.
+  handle: func(
```

Comment:
We will almost certainly want a way of passing request metadata here, for both "edge" metadata like the remote IP address and to aid component wrapper "middleware" use-cases like authentication.

- Go threads a separate general-purpose "[Context](https://pkg.go.dev/context#WithValue)" argument through call stacks
- Hyper packs metadata into the Request type with "[extensions](https://docs.rs/hyper/0.14.24/hyper/struct.Request.html#method.extensions)".

Reply:
If we're talking about generic data passed along with a request that isn't part of the HTTP-semantics-defined Fields resource, then I think this might not belong in the HTTP interface (where, in general, this generic data will be sheared off and lost as soon as the request travels over a real HTTP connection) but rather as a (perhaps middleware-specific) interface layered on top of HTTP, defining a new resource type that *contains* an HTTP request (and whatever other contextual metadata).  This also helps make it clear in the component's signature when a component is expecting to send/receive these annotated requests (to/from other components linked into the same middleware framework) vs. when a request is coming/going in/out via standard HTTP semantics.  At least, that's my first impression for how to address this sort of use case.

Reply:
That seems reasonable for middleware. What do you think about connection metadata like remote IP or peer TLS certs?

Reply:
Oops, right, I forgot to reply to that part of your last comment.  Potentially these make sense, but lets file separate issues to discuss these independently.  If you would like to do that, that'd be great.

Reply:
https://github.com/WebAssembly/wasi-http/issues/4